ql/src/java/org/apache/hadoop/hive/ql/plan/ExprDynamicParamDesc.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.ql.plan;
+
+import java.io.Serializable;
+import java.util.List;
+
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+import org.apache.hadoop.hive.common.StringInternUtils;
+import org.apache.hadoop.hive.serde.serdeConstants;
+import org.apache.hadoop.hive.serde2.objectinspector.ConstantObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector.Category;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorUtils;
+import org.apache.hadoop.hive.serde2.typeinfo.BaseCharTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.StructTypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfo;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoFactory;
+import org.apache.hadoop.hive.serde2.typeinfo.TypeInfoUtils;
+
+/**
+ * A constant expression.
+ */
+public class ExprDynamicParamDesc extends ExprNodeDesc implements Serializable {
+  private static final long serialVersionUID = 1L;
+  final protected transient static char[] hexArray = "0123456789ABCDEF".toCharArray();
+
+  private int index;
+  private Object value;
+
+  public ExprDynamicParamDesc() {
+  }
+
+  public ExprDynamicParamDesc(TypeInfo typeInfo, int index, Object value) {
+    super(typeInfo);
+    this.index =  index;
+    this.value = value;
+  }
+
+  public Object getValue() {
+    return value;
+  }
+
+  public int getIndex() {
+    return index;
+  }
+
+
+  @Override
+  public String toString() {
+    return "Dynamic Parameter " + " index: " + index;
```

Comment:
I was thinking we could use something more compact for the string representation in the plan -> `":" + index`. What do you think?

Reply:
"Dynamic Parameter" makes it clear that the expression in an explain plan is dynamic expression. Just showing index will make it hard to read.
What is the benefit of making it more compact?

Reply:
I just think that using `":" + index`, `"$" + index`, or any other simpler variants, will make the plans more readable rather than having the `dynamic parameter index: idx` interleaved in the middle of complex expressions. What do other RDBMSs show, e.g., Postgres or MySQL?
Maybe we could have different variants for user vs normal explain plans.

Reply:
Ok let me update it to `$<idx` for now. I am not sure what other databases use. I will try to figure out.